<h2 style='text-align: center;'> Hamoye Stage C Quiz Questions </h2>

In [1]:
import numpy as np
import pandas as pd

In [2]:
# reading the dataset to a dataframe
data = pd.read_csv('Data_for_UCI_named.csv')

In [3]:
data.head()

tau1      tau2      tau3      tau4        p1        p2        p3  \
0  2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1  9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2  8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3  0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4  3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   

         p4        g1        g2        g3        g4      stab     stabf  
0 -1.723086  0.650456  0.859578  0.887445  0.958034  0.055347  unstable  
1 -1.255012  0.413441  0.862414  0.562139  0.781760 -0.005957    stable  
2 -0.920492  0.163041  0.766689  0.839444  0.109853  0.003471  unstable  
3 -0.997374  0.446209  0.976744  0.929381  0.362718  0.028871  unstable  
4 -0.554305  0.797110  0.455450  0.656947  0.820923  0.049860  unstable

In [4]:
data.shape

(10000, 14)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   tau1    10000 non-null  float64
 1   tau2    10000 non-null  float64
 2   tau3    10000 non-null  float64
 3   tau4    10000 non-null  float64
 4   p1      10000 non-null  float64
 5   p2      10000 non-null  float64
 6   p3      10000 non-null  float64
 7   p4      10000 non-null  float64
 8   g1      10000 non-null  float64
 9   g2      10000 non-null  float64
 10  g3      10000 non-null  float64
 11  g4      10000 non-null  float64
 12  stab    10000 non-null  float64
 13  stabf   10000 non-null  object 
dtypes: float64(13), object(1)
memory usage: 1.1+ MB


In [6]:
data.duplicated().any()

False

In [7]:
data[ 'stabf' ].value_counts()

unstable    6380
stable      3620
Name: stabf, dtype: int64

In [8]:
# importing the necessar libraries and packages
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier
import xgboost as xgb
import lightgbm as lgb
from sklearn.metrics import accuracy_score

In [9]:
X = data.drop(['stab', 'stabf'], axis=1)
y = data['stabf']

# Split the data into train and test sets with a random state of 1
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

# Apply standard scaling to the train and test sets
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)
x_test_scaled = scaler.transform(X_test)

> Use label encoder to transform the 'stable' and 'unstable' variables in stabf column  to 0 and 1.

In [10]:
from sklearn.preprocessing import LabelEncoder

# Create a LabelEncoder object
label_encoder = LabelEncoder()

# Fit and transform the 'stabf' column in the train set
y_train_encoded = label_encoder.fit_transform(y_train)

# Transform the 'stabf' column in the test set
y_test_encoded = label_encoder.transform(y_test)

### Answering the quiz questions

> What is the accuracy on the test set using the random forest classifier? In 4 decimal places.

In [11]:
# Train a Random Forest Classifier
rf_clf = RandomForestClassifier(random_state=1)
rf_clf.fit(X_train_scaled, y_train_encoded)

# Make predictions on the test set
rf_pred = rf_clf.predict(x_test_scaled)

# Evaluate the accuracy of the model
rf_accuracy = accuracy_score(y_test_encoded, rf_pred)
print(f"Random Forest Accuracy: {rf_accuracy:.4f}")

Random Forest Accuracy: 0.9290


> What is the accuracy on the test set using the XGboost classifier? In 4 decimal places.

In [12]:
xgb_clf = xgb.XGBClassifier(random_state=1)
xgb_clf.fit(X_train_scaled, y_train_encoded)

# Make predictions on the test set
xgb_pred = xgb_clf.predict(x_test_scaled)

# Evaluate the accuracy of the model
xgb_accuracy = accuracy_score(y_test_encoded, xgb_pred)
print("XGBoost Accuracy:", xgb_accuracy)

XGBoost Accuracy: 0.9455


> What is the accuracy on the test set using the LGBM classifier? In 4 decimal places.

In [13]:
# Train a LightGBM model
lgb_clf = lgb.LGBMClassifier(random_state=1)
lgb_clf.fit(X_train_scaled, y_train_encoded)

# Make predictions on the test set
lgb_pred = lgb_clf.predict(x_test_scaled)

# Evaluate the accuracy of the model
lgb_accuracy = accuracy_score(y_test_encoded, lgb_pred)
print(f"LightGBM Accuracy: {lgb_accuracy:.4f}")
#print(f"Mean Absolute Error: {mae:.2f}")

LightGBM Accuracy: 0.9400


> Using the ExtraTreesClassifier as your estimator with cv=5, n_iter=10, scoring = 'accuracy', n_jobs = -1, verbose = 1 and random_state = 1. What are the best hyperparameters from the randomized search CV?

In [14]:
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.model_selection import RandomizedSearchCV

# Create the ExtraTreesClassifier estimator
estimator = ExtraTreesClassifier(random_state=1)

#combination of hyperparameters
n_estimators = [50, 100, 300, 500, 1000]

min_samples_split = [2, 3, 5, 7, 9]

min_samples_leaf = [1, 2, 4, 6, 8]

max_features = ['auto', 'sqrt', 'log2', None] 

# Define the hyperparameter grid for RandomizedSearchCV
hyperparameter_grid = {'n_estimators': n_estimators,

                       'min_samples_leaf': min_samples_leaf,

                       'min_samples_split': min_samples_split,

                       'max_features': max_features}


In [15]:
# Create the RandomizedSearchCV object
random_search = RandomizedSearchCV(estimator, hyperparameter_grid, cv=5, n_iter=10, scoring='accuracy',
                                   n_jobs=-1, verbose=1, random_state=1)

# Fit the RandomizedSearchCV object to the data
random_search.fit(X_train_scaled, y_train_encoded)

# Print the best hyperparameters found
print("Best Hyperparameters:", random_search.best_params_)

Fitting 5 folds for each of 10 candidates, totalling 50 fits
Best Hyperparameters: {'n_estimators': 1000, 'min_samples_split': 2, 'min_samples_leaf': 8, 'max_features': None}


> Train a new ExtraTreesClassifier Model with the new Hyperparameters from the RandomizedSearchCV (with random_state = 1). Is the accuracy of the new optimal model higher or lower than the initial ExtraTreesClassifier model with no hyperparameter tuning?

In [16]:
# Train an Extra Trees Classifier
et_clf = ExtraTreesClassifier(random_state=1)
et_clf.fit(X_train_scaled, y_train_encoded)

# Make predictions on the test set
et_pred = et_clf.predict(x_test_scaled)

# Evaluate the accuracy of the model
et_accuracy = accuracy_score(y_test_encoded, et_pred)
print("Extra Trees Accuracy:", et_accuracy)

Extra Trees Accuracy: 0.928


In [17]:
# Create the new ExtraTreesClassifier with the optimal hyperparameters
new_et_clf = ExtraTreesClassifier(random_state=1, **random_search.best_params_)

# Fit the new ExtraTreesClassifier on the training data
new_et_clf.fit(X_train_scaled, y_train_encoded)

# Make predictions on the test set using the new model
new_et_pred = new_et_clf.predict(x_test_scaled)

# Evaluate the accuracy of the new model
new_et_accuracy = accuracy_score(y_test_encoded, new_et_pred)
print("Accuracy of the New Optimal ExtraTreesClassifier Model:", new_et_accuracy)

Accuracy of the New Optimal ExtraTreesClassifier Model: 0.927


> The accuracy of the ExtraTreesClassifier without hyperparameter tuning is 0.928 (93%) while that of the classifier with hyperparameter tuning is 0.927 (93%), So we can say that there is no change in the accuracy of the model.

> Find the feature importance using the optimal ExtraTreesClassifier model. Which features are the most and least important respectively?

In [18]:
# Get the feature importances from the optimal ExtraTreesClassifier model
feature_importances = new_et_clf.feature_importances_

# Create a DataFrame to display the feature importances
importance_df = pd.DataFrame({'Feature': X.columns, 'Importance': feature_importances})

# Sort the DataFrame by importance in descending order
importance_df = importance_df.sort_values('Importance', ascending=False)

# Display the feature importances
print(importance_df)

# Print the feature with the most importance
print("Feature with the most importance:", importance_df.iloc[0]['Feature'])

# Print the feature with the highest weight
print("Feature with the least importance:", importance_df.iloc[-1]['Feature'])

   Feature  Importance
1     tau2    0.140508
0     tau1    0.137240
3     tau4    0.135417
2     tau3    0.134680
10      g3    0.113063
11      g4    0.109541
9       g2    0.107578
8       g1    0.102562
6       p3    0.005429
5       p2    0.005337
7       p4    0.004962
4       p1    0.003684
Feature with the most importance: tau2
Feature with the least importance: p1
